src/state/effects/sources.effect.ts

```diff
@@ -125,10 +125,18 @@ export function hookSourcesEffect() {
         case 'toggleSourcePin': {
           const [id, pinned] = args;
 
+          if (store.playing) {
+            store.setBufferPause(true);
+          }
+
           if (pinned) {
-            SourceHelper.pin(id);
+            await SourceHelper.pin(id);
           } else {
-            SourceHelper.unpin(id);
+            await SourceHelper.unpin(id);
+          }
+
+          if (store.playing) {
+            store.setBufferPause(false);
```

Comment:
This block loops finished sources when the user attempts to pin/unpin a source near the end of the timeline.

Reply:
The problem wasn't here.